src/hooks/useCreateMap.ts

```diff
@@ -0,0 +1,213 @@
+          // Handle successful uploads on complete
+          // For instance, get the download URL: https://firebasestorage.googleapis.com/...
+          const downloadURL = await getDownloadURL(uploadTask.snapshot.ref)
+          console.log('File available at', downloadURL)
+        },
+      )
+    },
+    [user, storage],
+  )
+
+  const createObservationsDocs = useCallback(
+    async (files: FileType[], mapPath: string) => {
+      const pointsJson = getJsonFromFiles(files, 'points.json') as PointsType
+      const images = getImagesFromFiles(files)
+      setTotalFiles(images.length)
+
+      totalBytesRef.current = images.reduce(
+        (acc, file) => (file.type === 'arraybuffer' ? acc + parseInt(file.data.byteLength) : acc),
+        0,
+      )
+
+      const points = pointsJson.features.map((feature) => {
+        const image = images.find((file) => path.basename(file.name) === feature.properties?.image)
+
+        if (!image && feature.properties?.image) {
+          console.log(`Missing image ${feature.properties?.image as string}`)
+          setError(new Error(`Missing image ${feature.properties.image as string}`))
+        }
+
+        return {
+          ...feature,
+          properties: { ...feature.properties, image: image ? image.hashedName : null },
+        }
+      })
+      const pointsWithIds = points.map((f) => ({
+        ...f,
+        properties: {
+          ...f.properties,
+          _id: md5(stringify(f)),
+        },
+      }))
+
+      const batch = writeBatch(db)
+
+      const observationsPath = `${mapPath}/observations`
+
+      pointsWithIds.forEach((point) => {
+        const pointRef = doc(db, observationsPath, point.properties._id)
+        batch.set(pointRef, point)
+      })
+
+      images.forEach((imageFile) => {
+        if (cancelRef.current) return // bail if component is unmounted
+        setCurrentFile((c) => c + 1)
+        try {
+          uploadImage(imageFile.hashedName || '', imageFile.data)
+        } catch (e) {
+          // Continue uploads after a failed upload, but mark as error
+          // setError(e)
+        }
+      })
+      await batch.commit()
+
+      console.log(`${observationsPath} set`)
+    },
+    [uploadImage],
+  )
+
+  const createMap = useCallback(
+    async (files: FileType[], id?: string) => {
+      setLoading(true)
+      setDone(false)
+      const mapPath = await createMapDoc(files)
+      await createObservationsDocs(files, mapPath)
+      setLoading(false)
+      setDone(true)
+    },
+    [createMapDoc, createObservationsDocs],
+  )
+
+  function updateProgress(bytesTransferred: number) {
+    if (!totalBytesRef.current) return
+    const transferred = bytesTransferredRef.current + bytesTransferred
+    setProgress(Math.ceil((transferred / totalBytesRef.current) * 100))
+  }
+
+  return {
+    createMap,
+    progress: {
+      currentFile,
+      completed: progress,
+      totalFiles,
+      error,
+      done,
+      // id,
+      loading,
```

Comment:
do we need `currentFile and total file` and `completed`

Reply:
`currentFile` is the index of the file being uploaded, `totalFiles` is the number of files and `completed` is the total % completed upload